docs/guide/rules-concept.md

```diff
@@ -0,0 +1,181 @@
+### Validation rule handlers
+
+#### Creating your own validation rule handlers
+
+##### Basic usage
+
+To create your own validation rule handler you should implement `RuleHandlerInterface`:
+
+```php
+namespace MyVendor\Rules;
+
+use Yiisoft\Validator\DataSetInterface;
+use Yiisoft\Validator\Exception\UnexpectedRuleException;use Yiisoft\Validator\FormatterInterface;use Yiisoft\Validator\Result;
+use Yiisoft\Validator\RuleHandlerInterface;use Yiisoft\Validator\RuleInterface;
+
+final class PiHandler implements RuleHandlerInterface
+{
+    use FormatMessageTrait;
+    
+    private FormatterInterface $formatter;
+    
+    public function __construct(
+        ?FormatterInterface $formatter = null,
+    ) {
+        $this->formatter = $this->createFormatter();
+    }
+    
+    public function validate(mixed $value, object $rule, ?ValidationContext $context = null): Result
+    {
+        if (!$rule instanceof Pi) {
+            throw new UnexpectedRuleException(Pi::class, $rule);
+        }
+        
+        $result = new Result();
+        $equal = \abs($value - M_PI) < PHP_FLOAT_EPSILON;
+
+        if (!$equal) {
+            $result->addError($this->formatter->format('Value is not PI.'));
+        }
+
+        return $result;
+    }
+    
+    private function createFormatter(): FormatterInterface
+    {
+        // More complex logic
+        // ...
+        return CustomFormatter();
+    }
+}
+```
+
+Note that third argument in `validate()` is an instance of `ValidationContext` so you can use it if you need
+whole data set context. For example, implementation might be the following if you need to validate "company"
+property only if "hasCompany" is true:
+
+```php
+namespace MyVendor\Rules;
+
+use Yiisoft\Validator\DataSetInterface;
+use Yiisoft\Validator\Result;
+use Yiisoft\Validator\Rule;
+use Yiisoft\Validator\ValidationContext;
+
+final class CompanyNameHandler implements Rule\RuleHandlerInterface
+{
+    use FormatMessageTrait;
+    
+    private FormatterInterface $formatter;
+
+    public function validate(mixed $value, object $rule, ?ValidationContext $context = null): Result
+    {
+        if (!$rule instanceof CompanyName) {
+            throw new UnexpectedRuleException(CompanyName::class, $rule);
+        }
+        
+        $result = new Result();
+        $dataSet = $context->getDataSet();
+        $hasCompany = $dataSet->getAttributeValue('hasCompany') === true;
+
+        if ($hasCompany && $this->isCompanyNameValid($value) === false) {
+            $result->addError('Company name is not valid.');
+        }
+
+        return $result;
+    }
+
+    private function isCompanyNameValid(string $value): bool
+    {
+        // check company name    
+    }
+}
+```
+
+> Note: Do not call handler's `validate()` method directly. It must be used via Validator only.
+
+##### Resolving rule handler dependencies
+
+Basically, you can use `SimpleRuleHandlerResolver` to resolve rule handler.
```

Comment:
```suggestion
You can use `SimpleRuleHandlerResolver` to resolve rule handler.
```

Reply:
Will be fixed after finishing "Rules handlers resolvers" section.

Reply:
Suggest to add this section in a separate PR.